server/src/main/java/com/objectcomputing/checkins/security/ImpersonationController.java

```diff
@@ -0,0 +1,151 @@
+            if (auth.isPresent() && auth.get().getAttributes().get("email") != null) {
+                final Cookie jwt = request.getCookies().get(JWT);
+                if (jwt == null) {
+                    // The user is required to be logged in.  If this is null,
+                    // we are in an impossible state!
+                    LOG.error("Unable to locate the JWT");
+                } else {
+                    UsernamePasswordCredentials usernamePasswordCredentials = new UsernamePasswordCredentials(email, "");
+                    Flux<AuthenticationResponse> authenticationResponseFlux =
+                            Flux.from(authenticator.authenticate(request, usernamePasswordCredentials));
+                    return authenticationResponseFlux.map(authenticationResponse -> {
+                        if (authenticationResponse.isAuthenticated() && authenticationResponse.getAuthentication().isPresent()) {
+                            Authentication authentication = authenticationResponse.getAuthentication().get();
+                            // Get member profile by work email
+                            MemberProfile memberProfile = currentUserServices.findOrSaveUser("", "", email);
+                            String firstName = memberProfile.getFirstName() != null ? memberProfile.getFirstName() : "";
+                            String lastName = memberProfile.getLastName() != null ? memberProfile.getLastName() : "";
+
+                            Map<String, Object> newAttributes = new HashMap<>(authentication.getAttributes());
+                            newAttributes.put("email", memberProfile.getWorkEmail());
+                            newAttributes.put("name", firstName + ' ' + lastName);
+                            newAttributes.put("picture", "");
+                            Authentication updatedAuth = Authentication.build(authentication.getName(), authentication.getRoles(), newAttributes);
+
+                            eventPublisher.publishEvent(new LoginSuccessfulEvent(updatedAuth, null, Locale.getDefault()));
+                            // Store the old JWT to allow the user to revert the impersonation.
+                            return ((MutableHttpResponse)loginHandler.loginSuccess(updatedAuth, request)).cookie(
+                                       new NettyCookie(originalJWT, jwt.getValue()).path("/").sameSite(SameSite.Strict)
+                                       .maxAge(jwt.getMaxAge()));
+                        } else {
+                            eventPublisher.publishEvent(new LoginFailedEvent(authenticationResponse, null, null, Locale.getDefault()));
+                            return loginHandler.loginFailed(authenticationResponse, request);
+                        }
+                    }).single(Mono.just(HttpResponse.status(HttpStatus.UNAUTHORIZED)));
+                }
+            } else {
+                LOG.error("Attempted impersonation without authentication.");
```

Comment:
`HttpResponse.serverError()` here...

Reply:
You can still do these though...just `return Mono.just(HttpResponse.serverError());`